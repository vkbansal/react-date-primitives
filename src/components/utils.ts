import { DayOfMonthSubset } from './types';

const monthsCache: Map<string, DayOfMonthSubset[][]> = new Map();

export function setDay(date: Date, day: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(day);

    return newDate;
}

export function addDays(date: Date, days: number): Date {
    return setDay(date, date.getDate() + days);
}

export function setMonth(date: Date, month: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setMonth(month);

    return newDate;
}

export function addMonths(date: Date, months: number): Date {
    return setMonth(date, date.getMonth() + months);
}

export function setYear(date: Date, year: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(year);

    return newDate;
}

export function startOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(1);
    newDate.setHours(0, 0, 0, 0);

    return newDate;
}

export function endOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(date.getFullYear(), date.getMonth() + 1, 0);
    newDate.setHours(23, 59, 59, 999);

    return newDate;
}

export function isSameMonth(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth()
    );
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth() &&
        dateLeft.getDate() === dateRight.getDate()
    );
}

export function isDayAfter(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() > newDateToCompare.getTime();
}

export function isDayBefore(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() < newDateToCompare.getTime();
}

export function callIfExists(callback?: any, ...args: any[]) {
    if (typeof callback === 'function') {
        callback(...args);
    }
}

export function getDatesofMonth(month: Date): DayOfMonthSubset[][] {
    const firstDate = startOfMonth(month);
    const key = `${firstDate.getFullYear()}-${firstDate.getMonth()}`;

    const data = monthsCache.get(key);

    if (data) {
        return data;
    }

    const firstDayInCurrentMonth = firstDate.getDay();

    const newdata: DayOfMonthSubset[][] = Array.from(
        { length: 6 },
        (_1, i): DayOfMonthSubset[] => {
            const week = Array.from(
                { length: 7 },
                (_2, j): DayOfMonthSubset => {
                    const day = addDays(firstDate, i * 7 + j - firstDayInCurrentMonth);

                    return { date: day, inCurrentMonth: isSameMonth(firstDate, day) };
                }
            );

            return week;
        }
    );

    monthsCache.set(key, newdata);

    return newdata;
}
