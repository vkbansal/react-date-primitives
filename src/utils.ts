export interface DayOfMonth {
    date: Date;
    inCurrentMonth: boolean;
}

export type DaysOfMonth = DayOfMonth[][];

const monthsCache: Map<string, DaysOfMonth> = new Map();

export function setDay(date: Date, day: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(day);

    return newDate;
}

export function addDays(date: Date, days: number): Date {
    return setDay(date, date.getDate() + days);
}

export function setMonth(date: Date, month: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setMonth(month);

    return newDate;
}

export function addMonths(date: Date, months: number): Date {
    return setMonth(date, date.getMonth() + months);
}

export function setYear(date: Date, year: number): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(year);

    return newDate;
}

export function startOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setDate(1);
    newDate.setHours(0, 0, 0, 0);

    return newDate;
}

export function endOfMonth(date: Date): Date {
    const newDate = new Date(date.getTime());

    newDate.setFullYear(date.getFullYear(), date.getMonth() + 1, 0);
    newDate.setHours(23, 59, 59, 999);

    return newDate;
}

export function isSameMonth(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth()
    );
}

export function isSameDay(dateLeft: Date, dateRight: Date): boolean {
    return (
        dateLeft.getFullYear() === dateRight.getFullYear() &&
        dateLeft.getMonth() === dateRight.getMonth() &&
        dateLeft.getDate() === dateRight.getDate()
    );
}

export function isDayAfter(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() > newDateToCompare.getTime();
}

export function isDayBefore(date: Date, dateToCompare: Date): boolean {
    const newDate = new Date(date.getTime());
    const newDateToCompare = new Date(dateToCompare.getTime());

    newDate.setHours(0, 0, 0, 0);
    newDateToCompare.setHours(0, 0, 0, 0);

    return newDate.getTime() < newDateToCompare.getTime();
}

export function callIfExists(callback?: any, ...args: any[]) {
    if (typeof callback === 'function') {
        callback(...args);
    }
}

export function getDaysOfMonth(month: Date): DayOfMonth[][] {
    const firstDate = startOfMonth(month);
    const key = `${firstDate.getFullYear()}-${firstDate.getMonth()}`;

    const data = monthsCache.get(key);

    if (data) {
        return data;
    }

    const firstDayInCurrentMonth = firstDate.getDay();

    const newdata: DaysOfMonth = Array.from(
        { length: 6 },
        (_1, i): DayOfMonth[] => {
            const week = Array.from(
                { length: 7 },
                (_2, j): DayOfMonth => {
                    const day = addDays(firstDate, i * 7 + j - firstDayInCurrentMonth);

                    return { date: day, inCurrentMonth: isSameMonth(firstDate, day) };
                }
            );

            return week;
        }
    );

    monthsCache.set(key, newdata);

    return newdata;
}
