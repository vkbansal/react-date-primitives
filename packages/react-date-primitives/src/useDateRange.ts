import { useState, useCallback } from 'react';
import type { Dispatch, SetStateAction } from 'react';

import { getDaysOfRangeMonth, isDayBefore, DayName, RangeMonths } from './utils';

export type DateRange = [Date | null, Date | null];

export interface UseDateRangeOptions {
  range?: [Date | null, Date | null];
  rangeEndDate?: Date;
  weekStartsOn?: DayName;
}

export interface UseDateRangeReturn extends RangeMonths {
  readonly range: [Date | null, Date | null];
  setMonths: Dispatch<SetStateAction<Date[]>>;
  setRange: Dispatch<SetStateAction<[Date | null, Date | null]>>;
  setStartOfWeek: Dispatch<SetStateAction<DayName>>;
}

export function useDateRange(
  rangeMonths: Date[],
  options: UseDateRangeOptions = {}
): UseDateRangeReturn {
  const [currentMonths, setCurrentMonths] = useState<Date[]>(rangeMonths);
  const [startOfWeek, setStartOfWeek] = useState(options?.weekStartsOn || DayName.SUNDAY);
  const [range, setRange] = useState<DateRange>([
    options?.range?.[0] || null,
    options?.range?.[1] || null
  ]);

  const { months, daysOfWeek } = getDaysOfRangeMonth(
    currentMonths,
    range[0],
    range[1],
    startOfWeek
  );

  const setRangeCallBack = useCallback(
    (newRange: SetStateAction<DateRange>) => {
      setRange((oldRange) => {
        const [start, end] = typeof newRange === 'function' ? newRange(oldRange) : newRange;
        if (start && end && isDayBefore(end, start)) {
          throw new RangeError(`Range end cannot be before the range start`);
        }

        return [start, end];
      });
    },
    [setRange]
  );

  return {
    months,
    daysOfWeek,
    range,
    setMonths: setCurrentMonths,
    setRange: setRangeCallBack,
    setStartOfWeek
  };
}
