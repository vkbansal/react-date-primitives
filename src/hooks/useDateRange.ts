import { useState } from 'react';

import { DayOfMonth, isDayBefore, isDayAfter, isSameDay } from '../utils';

export interface DayOfRangeMonth extends DayOfMonth {
    inRange: boolean;
    selected: boolean;
    isStart: boolean;
    isEnd: boolean;
}

export interface UseDateRange {
    handleDayClick(day: Date): void;
    handleDayHover(day: Date): void;
    startDate: Date;
    endDate?: Date;
    processMonth(month: DayOfMonth[][]): DayOfRangeMonth[][];
}

export function useDateRange(
    initialStartDate: Date = new Date(),
    initialEndDate?: Date
): UseDateRange {
    const [startDate, setStartDate] = useState(initialStartDate);
    const [endDate, setEndDate] = useState(initialEndDate);
    const [selectionActive, setSelectionActive] = useState(false);

    function handleDayClick(day: Date) {
        if (startDate && selectionActive && !isDayBefore(day, startDate)) {
            setStartDate(startDate);
            setEndDate(day);
            setSelectionActive(false);
        } else {
            setStartDate(day);
            setEndDate(undefined);
            setSelectionActive(true);
        }
    }

    function handleDayHover(day: Date) {
        if (selectionActive && startDate && !isDayBefore(day, startDate)) {
            setEndDate(day);
        }
    }

    function processMonth(month: DayOfMonth[][]): DayOfRangeMonth[][] {
        return month.map<DayOfRangeMonth[]>((week) =>
            week.map<DayOfRangeMonth>((day) => {
                const isStart = startDate && isSameDay(day.date, startDate);
                const isEnd = Boolean(endDate && isSameDay(day.date, endDate));

                return {
                    ...day,
                    selected: isStart || isEnd,
                    inRange: Boolean(
                        startDate &&
                            isDayAfter(day.date, startDate) &&
                            endDate &&
                            isDayBefore(day.date, endDate)
                    ),
                    isStart,
                    isEnd
                };
            })
        );
    }

    return { handleDayClick, handleDayHover, startDate, endDate, processMonth };
}
