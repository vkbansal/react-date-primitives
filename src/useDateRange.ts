import { useState, SetStateAction, Dispatch } from 'react';

import { getDaysOfRangeMonth, isSameDay, isDayAfter, DayOfWeek, RangeMonths } from './utils';

export interface UseDateRangeOptions {
    rangeStartDate?: Date;
    rangeEndDate?: Date;
    weekStartsOn?: DayOfWeek;
}

export interface DateRange extends RangeMonths {
    readonly startDate: Date | null;
    readonly endDate: Date | null;
    setMonths: Dispatch<SetStateAction<Date[]>>;
    setStartDate: Dispatch<SetStateAction<Date | null>>;
    setEndDate: Dispatch<SetStateAction<Date | null>>;
}

export function useDateRange(rangeMonths: Date[], options: UseDateRangeOptions = {}): DateRange {
    const [currentMonths, setCurrentMonths] = useState<Date[]>(rangeMonths);
    const [startDate, setStartDate] = useState<Date | null>(options.rangeEndDate || null);
    const [endDate, setEndDate] = useState<Date | null>(options.rangeEndDate || null);

    const { months, daysOfWeek } = getDaysOfRangeMonth(currentMonths, startDate, endDate);

    return {
        months,
        daysOfWeek,
        startDate,
        endDate,
        setMonths: setCurrentMonths,
        setStartDate(date: SetStateAction<Date | null>): void {
            setStartDate(date);
            setEndDate(null);
        },
        setEndDate(date: SetStateAction<Date | null>): void {
            if (
                startDate &&
                date instanceof Date &&
                (isSameDay(date, startDate) || isDayAfter(date, startDate))
            ) {
                setEndDate(date);
            } else if (typeof date === 'function') {
                setEndDate(date);
            }
        }
    };
}
