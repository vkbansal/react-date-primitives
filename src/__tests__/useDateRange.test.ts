import { renderHook, act } from '@testing-library/react-hooks';

import { useDateRange } from '../useDateRange';
import { isSameDay, addMonths, Day } from '../utils';

import DateSerializer from './DateSerializer';
import RangeMonthsSerializer from './RangeMonthsSerializer';

expect.addSnapshotSerializer(DateSerializer);
expect.addSnapshotSerializer(RangeMonthsSerializer);

describe('useDateRange hook tests', () => {
    const month = new Date(2020, 0 /* Jan */, 1, 0, 0, 0, 0);
    const months = Array.from({ length: 12 }, (_, i) => addMonths(month, i));

    test('returns basic properties', () => {
        const { result } = renderHook(() => useDateRange(months.slice(0, 3)));

        expect(result.current.startDate).toBeNull();
        expect(result.current.endDate).toBeNull();
        expect(Array.isArray(result.current.months)).toBe(true);
        expect(result.current.months.every((month) => Array.isArray(month.days))).toBe(true);
        expect(
            result.current.months.every((month) => month.days.every((week) => Array.isArray(week)))
        ).toBe(true);
        expect(result.current.months[0].days[0][0]).toMatchSnapshot('Single Day');
        expect(typeof result.current.setStartDate).toBe('function');
        expect(typeof result.current.setEndDate).toBe('function');
        expect(typeof result.current.setMonths).toBe('function');
    });

    describe('setEndDate', () => {
        describe('`Date` value', () => {
            test('works for only when endDate > startDate', () => {
                const date1 = new Date(2020, 0 /* Jan */, 1, 0, 0, 0, 0);
                const date2 = new Date(2020, 0 /* Jan */, 9, 0, 0, 0, 0);
                const { result } = renderHook(() => useDateRange(months.slice(0, 3)));

                expect(result.current.startDate).toBeNull();
                expect(result.current.endDate).toBeNull();
                expect(result.current).toMatchSnapshot('01. No Selection');

                // set a startDate
                act(() => {
                    result.current.setStartDate(date1);
                });
                expect(isSameDay(result.current.startDate!, date1)).toBe(true);
                expect(result.current.endDate).toBeNull();

                expect(result.current).toMatchSnapshot('02. Start date selected');

                // set an endDate > startDate
                act(() => {
                    result.current.setEndDate(date2);
                });
                expect(isSameDay(result.current.startDate!, date1)).toBe(true);
                expect(isSameDay(result.current.endDate!, date2)).toBe(true);

                expect(result.current).toMatchSnapshot('03. Start & End dates selected');

                // set a startDate
                act(() => {
                    result.current.setStartDate(date2);
                });
                expect(isSameDay(result.current.startDate!, date2)).toBe(true);
                expect(result.current.endDate).toBeNull();

                expect(result.current).toMatchSnapshot('04. Start date selected again');

                // set an endDate < startDate
                act(() => {
                    result.current.setEndDate(date1);
                });
                expect(isSameDay(result.current.startDate!, date2)).toBe(true);
                expect(result.current.endDate).toBeNull();

                expect(result.current).toMatchSnapshot('05. Try selecting smaller end date again');
            });
        });

        describe('functional value', () => {
            test('works without any checks', () => {
                const date1 = new Date(2020, 0 /* Jan */, 1, 0, 0, 0, 0);
                const date2 = new Date(2020, 0 /* Jan */, 5, 0, 0, 0, 0);
                const date3 = new Date(2020, 0 /* Jan */, 9, 0, 0, 0, 0);
                const { result } = renderHook(() => useDateRange(months.slice(0, 3)));

                expect(result.current.startDate).toBeNull();
                expect(result.current.endDate).toBeNull();
                expect(result.current).toMatchSnapshot('01. No Selection');

                // set a startDate
                act(() => {
                    result.current.setStartDate(() => date1);
                });
                expect(isSameDay(result.current.startDate!, date1)).toBe(true);
                expect(result.current.endDate).toBeNull();

                expect(result.current).toMatchSnapshot('02. Start date selected');

                act(() => {
                    result.current.setEndDate(() => date3);
                });
                expect(isSameDay(result.current.startDate!, date1)).toBe(true);
                expect(isSameDay(result.current.endDate!, date3)).toBe(true);

                expect(result.current).toMatchSnapshot('03. Start & End dates selected');

                // set a endDate
                act(() => {
                    result.current.setEndDate(() => date2);
                });
                expect(isSameDay(result.current.endDate!, date2)).toBe(true);

                expect(result.current).toMatchSnapshot('04. End date selected again');

                // set an endDate > startDate
                act(() => {
                    result.current.setStartDate(() => date3);
                    result.current.setEndDate(() => date1);
                });
                expect(isSameDay(result.current.startDate!, date3)).toBe(true);
                expect(isSameDay(result.current.endDate!, date1)).toBe(true);

                expect(result.current).toMatchSnapshot('05. Try selecting smaller end date');
            });
        });
    });

    test('setMonths works', () => {
        const { result } = renderHook(() => useDateRange(months.slice(0, 3)));

        expect(result.current).toMatchSnapshot();

        act(() => {
            result.current.setMonths(months.slice(3, 6));
        });

        expect(result.current).toMatchSnapshot();
    });

    test('changing `weekStartsOn` works', () => {
        const { result } = renderHook(() => useDateRange(months.slice(0, 3)));
        expect(result.current.daysOfWeek[0]).toBe(Day.SUNDAY);
        expect(result.current).toMatchSnapshot(Day.SUNDAY);

        act(() => {
            result.current.setStartOfWeek(Day.MONDAY);
        });

        expect(result.current.daysOfWeek[0]).toBe(Day.MONDAY);
        expect(result.current).toMatchSnapshot(Day.MONDAY);
    });
});
