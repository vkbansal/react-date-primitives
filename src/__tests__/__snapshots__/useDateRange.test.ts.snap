// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDateRange hook tests returns basic properties: Single Day 1`] = `
Object {
  "ISODateString": "2019-12-29",
  "__type": Symbol(DayOfRangeMonth),
  "date": 2019-12-29,
  "dayName": "SUNDAY",
  "inCurrentMonth": false,
  "inRange": false,
  "isEnd": false,
  "isStart": false,
}
`;

exports[`useDateRange hook tests selection works for only when endDate > startDate: 01. No Selection 1`] = `
         S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S
Jan =>     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |    
Feb =>     |     |     |     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |     |     |     |     |     |     |    
Mar =>  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |     |     |     |    
`;

exports[`useDateRange hook tests selection works for only when endDate > startDate: 02. Start date selected 1`] = `
         S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S
Jan =>     |     |     | *01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |    
Feb =>     |     |     |     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |     |     |     |     |     |     |    
Mar =>  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |     |     |     |    
`;

exports[`useDateRange hook tests selection works for only when endDate > startDate: 03. Start & End dates selected 1`] = `
         S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S
Jan =>     |     |     | *01 | *02 | *03 | *04 | *05 | *06 | *07 | *08 | *09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |    
Feb =>     |     |     |     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |     |     |     |     |     |     |    
Mar =>  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |     |     |     |    
`;

exports[`useDateRange hook tests selection works for only when endDate > startDate: 04. Start date selected again 1`] = `
         S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S
Jan =>     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 | *09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |    
Feb =>     |     |     |     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |     |     |     |     |     |     |    
Mar =>  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |     |     |     |    
`;

exports[`useDateRange hook tests selection works for only when endDate > startDate: 05. Try selecting smaller end date again 1`] = `
         S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S
Jan =>     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 | *09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |    
Feb =>     |     |     |     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |     |     |     |     |     |     |    
Mar =>  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |     |     |     |    
`;

exports[`useDateRange hook tests setMonths works 1`] = `
         S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S
Jan =>     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |    
Feb =>     |     |     |     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |     |     |     |     |     |     |    
Mar =>  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |     |     |     |     |     |    
`;

exports[`useDateRange hook tests setMonths works 2`] = `
         S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S |   S |   M |   T |   W |   T |   F |   S
Apr =>     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |     |     |     |     |     |     |     |     |    
May =>     |     |     |     |     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |  31 |     |     |     |     |     |    
Jun =>     |  01 |  02 |  03 |  04 |  05 |  06 |  07 |  08 |  09 |  10 |  11 |  12 |  13 |  14 |  15 |  16 |  17 |  18 |  19 |  20 |  21 |  22 |  23 |  24 |  25 |  26 |  27 |  28 |  29 |  30 |     |     |     |     |     |     |     |     |     |     |    
`;
